### PR TITLE
Add support for nodeName different from  machine name on GCE

### PR DESCRIPTION
**Pull request by enxebre**
**What this PR does / why we need it**:
Maps fqdn hostname to instance name on gce

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
https://github.com/kubernetes/kubernetes/issues/54502

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by dims** [comment]
Don't think so @enxebre

**Comment by enxebre** [comment]
cool green now, any thoughts cc @cheftako ?

**Comment by enxebre** [comment]
any feedback on this? cc @bowei

File: pkg/cloudprovider/providers/gce/gce_util.go [hunk 1]
```diff
@@ -120,12 +123,47 @@ func lastComponent(s string) string {
...
+	return getNodeName(instance.Name, currentInstanceHostname, currentInstanceProjectId)
+}
+
+func getNodeName(instanceName string, instanceHostname string, instanceProjectId string) (types.NodeName, error) {
+	matched, err := regexp.MatchString(".*\\.c\\."+instanceProjectId+"\\.internal", instanceHostname)
+	if err != nil {
+		return "", err
+	}
+	if matched {
+		return types.NodeName(instanceName + ".c." + instanceProjectId + ".internal"), nil
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_util.go]: I'm not sure if this is a great idea. The naming scheme used here `instance.c.project-id.internal` is the DNS name of the instance and is not guaranteed to stay the same, and there could be multiple names pointing to the same instance...

**Comment by enxebre** [comment]
@bowei I agree that baking the FQDN is not ideal. Added a more ambiguous approach by getting nodeList now. Any thoughts?

**Comment by dims** [comment]
@cheftako @bowei @pmorie ping

**Pull request closed without merging**